### PR TITLE
Add LazyRootFolder

### PR DESCRIPTION
**Pull request by rullzer**
To make https://github.com/owncloud/core/pull/23919 a bit smaller.

This adds a LazyRootFolder. That will only initialize the moment a function is called. This allows the share Manager/Providers/Shares to be a lot more lazy.

This lazyness is required if we want to use the shareManager also in the sharedMount. Else we get cyclic dependencies.

CC: @icewind1991 @DeepDiver1975 @nickvergessen @PVince81 @MorrisJobke @LukasReschke

### CODE REVIEW

File: lib/private/files/node/lazyroot.php [hunk 1]
```diff
@@ -0,0 +1,456 @@
+<?php
+
```
Comment (nickvergessen) [inline lib/private/files/node/lazyroot.php]: missing license header

**Comment by nickvergessen** [comment]
Should we always use the lazy root?

**Comment by nickvergessen** [comment]
Looks good otherwise

**Comment by rullzer** [comment]
It might make sense to use it in more places eventually.

File: lib/private/files/node/lazyroot.php (hunk 1 above)
Comment (rullzer) [inline lib/private/files/node/lazyroot.php]: fixed

File: lib/private/files/node/lazyroot.php [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+		}
+
+		return call_user_func_array([$this->rootFolder, $method], $args);
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getUser() {
+		return $this->__call(__FUNCTION__, func_get_args());
```
Comment (icewind1991) [inline lib/private/files/node/lazyroot.php]: Is this needed? I would expect the magic `_call` properties to do this automatically

**Comment by icewind1991** [comment]
:+1: otherwise

File: lib/private/files/node/lazyroot.php (hunk 2 above)
Comment (rullzer) [inline lib/private/files/node/lazyroot.php]: It is needed since it implements the interface.

Comment (nickvergessen) [inline lib/private/files/node/lazyroot.php]: Well that's odd, but okay then

Comment (rullzer) [inline lib/private/files/node/lazyroot.php]: Let me double check... just to be sure...

Comment (rullzer) [inline lib/private/files/node/lazyroot.php]: yep all breaks down if they are removed.

**Comment by rullzer** [comment]
@nickvergessen can I cound that as a :+1: ?

**Comment by nickvergessen** [comment]
Sicne sharing and uploading/downloading still works, :+1:

**Pull request merged**